ruby_programming/automated_testing/test_driven_development.md

```diff
@@ -0,0 +1,206 @@
+
+~~~ bash
+mkdir tdd-lesson
+cd tdd-lesson
+mkdir lib
+mkdir spec
+touch lib/square.rb
+touch spec/square_spec.rb
+code .
+~~~
+
+Now, how might you want to TDD the creation of this class? Well, you know that you need to `describe` a class called `Square`, and that you'll need to describe its `#area` method, so start with setting all that up in a `square_spec.rb` file:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+
+  end
+end
+~~~
+
+Since `Square` will be initialized with a side length, do that initialization in a couple of different `context` blocks so that you can initialize differently sized squares to test the method:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+    context 'when side length is 4' do
+      let(:size_4_square) { described_class.new(4) }
+    end
+
+    context 'when side length is 6' do
+      let(:size_6_square) { described_class.new(6) }
+    end
+  end
+end
+~~~
+
+Next add an `it` block to each of the contexts to actually test that the method works as you expect:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#area" do
+    context 'when side length is 4' do
+      let(:size_4_square) { described_class.new(4) }
+
+      it ‘returns 16’ do
+        expect(size_4_square.area).to eq(16)
+      end
+    end
+
+    context 'when side length is 6' do
+      let(:size_6_square) { described_class.new(6) }
+
+      it ‘returns 36’ do
+        expect(size_6_square.area).to eq(36)
+      end
+    end
+  end
+end
+~~~
+
+Alright, that should have you set up to test `#area`, now run the tests by entering `rspec` in the terminal!
+
+The result should be a *red* failure/error, since you haven’t implemented `Square` at all yet. This is the `red` stage of the red-green-refactor cycle mentioned earlier, so the next step is making it `green` by implementing the expected functionality! Do that in our `lib/square.rb` file now:
+
+~~~rb
+class Square
+  def initialize(side_length)
+    @side_length = side_length
+  end
+
+  def area
+    @side_length * @side_length
+  end
+end
+~~~
+
+And now if you run the test suite again you should get a nice *green* output indicating no failures! Try it by entering `rspec` in the terminal again.
+
+Success! You are now on the green stage of the red-green-refactor cycle! Now you can refactor knowing that you can use the test to make sure to keep the existing functionality that you’ve built, maybe do:
+
+~~~rb
+class Square
+  def initialize(side_length)
+    @side_length = side_length
+  end
+
+  def area
+    @side_length ** 2
+  end
+end
+~~~
+
+Yeah, that’s a little cleaner! And you can run `rspec` again to make sure the tests still pass.
+
+Looks like you’re good! Now do the `#perimeter` method! First write the test:
+
+~~~rb
+require_relative '../lib/square'
+
+describe Square do
+  describe "#perimeter" do
+    context 'when side length is 4' do
```

Comment:
I would add a comment like `# omitted for brevity` with some blank lines between the two describe blocks (like the new RSpec Part 1 Basics Lesson)